### DISCUSSION TITLE
What is the best way to deal with HTTP 429 - Throttled - requests with PnP ?

### DISCUSSION THREAD

**Discussion by MartinHatch** [pnp/powershell#62] [General] [answered]
I have a script where I iterate through 22k items in a CSV file and update document sets for each item. 

So I'm basically repeatedly calling 

`Set-PnPListItem -List $listName -Identity $listItemId  -Values $itemValues`

However, every now and then the operation fails with HTTP 429 - Too Many Requests.

Is there a standard way of capturing this and identifying the Wait/Retry time which I understand would be in the response header.

**Comment by erwinvanhunen** [comment, marked as answer]
In principle there is nothing you need to do to handle 429s. PnP PowerShell makes use of the PnP Framework behind the scenes where we override way CSOM work with it comes to executing requests. We implemented an incremental backoff flow which means that if a 429 comes in we initially wait 0.5 seconds. Then we try again. If it still fails, we wait 2x0.5 seconds, then 2, then 4, then 8, then 16, 32, 64, 128, 256 seconds. We try up to a maximum of 10 times before giving up. That means that maximum wait time before we give up is 512 seconds, meaning more than 8 minutes. This delay has shown to handle the throttling errors is almost every case. We have seen that this covers throttling nicely.

Do you by any change in your script run other cmdlets not coming from PnP PowerShell, as all PnP PowerShell cmdlets use the throttling handling mechanism I just described?

**Reply by MartinHatch** [reply]
> ... We implemented an incremental backoff flow which means that if a 429 comes in we initially wait 0.5 seconds. Then we try again. If it still fails, we wait 2x0.5 seconds, then 2, then 4, then 8, then 16, 32, 64, 128, 256 seconds. We try up to a maximum of 10 times before giving up. That means that maximum wait time before we give up is 512 seconds, meaning more than 8 minutes. This delay has shown to handle the throttling errors is almost every case. We have seen that this covers throttling nicely.
> 

Query - why not use the response header value which tells you *exactly* how long to wait for?

**Reply by erwinvanhunen** [reply]
That's a long story... The short story is that it doesn't always tell you exactly how long to wait for the last time we checked.

**Reply by erwinvanhunen** [reply]
We did an intermediate release available as a nightly release, available tomorrow as version 0.3.13-nightly (notice: this will only available with PnP.PowerShell, not SharePointPnPPowerShellOnline). In that build we allow you to override the behavior of throttling handling. We're curious to see if it has any effects on your tenant. Use `Connect-PnPOnline <all your other parameters> -UseRetryAfterHeader`.

**Reply by erwinvanhunen** [reply]
Another short update: we did some additional tests here and we are strongly considering actually switching our approach over to using the retry-after header value.

**Reply by erwinvanhunen** [reply]
If we do that the -UseRetryAfterHeader property will be removed again from Connect-PnPOnline. Nevertheless I would be interested to see the tracelog from your end using that property with 0.3.13-nightly.

**Reply by erwinvanhunen** [reply]
You'll notice a false positive in the tracelog : duplicate log entries in the case of throttling. The second line will show the actual value reported back by the retry-after header, the first line shows our initial delay time, wrongly stated as seconds where it actually is milliseconds. We obviously will fix that in the next release.

**Reply by erwinvanhunen** [reply]
We decided to flip over to using the retry-after header value in general. I removed the switch from PowerShell again as it is now the default behavior. So 0.3.13-night contains the -UseRetryAfterHeader, 0.3.14-night and later will not contain the switch.

**Reply by pbjorklund** [reply]
Is there some way to write throttling information to logs/stdout? Edit: I see that you answered pretty much what I wanted in the other thread on this page.

**Comment by MartinHatch** [comment]
Literally just "Set-PnPListItem" over and over .. 

I randomly get 429 errors ...

**Comment by erwinvanhunen** [comment]
How do you connect to your tenant (e.g. what connection method/parameters do you use)? And which version of PnP PowerShell do you use? The 'legacy' version (SharePointPnPPowerShellOnline) or the new PnP.PowerShell version?

**Reply by MartinHatch** [reply]
The "legacy" version .. SharePointPnPPowerShellOnline

Using the July 2020 Release build for other backward compatibility reasons - but was also seeing the same error in the December 2020 build

**Reply by erwinvanhunen** [reply]
the reason I ask this is that if you use -PnPManagementShell on the old version or you use -Credentials with the new version you basically use the single PnP Management Shell Azure AD App registration/identity in your tenant towards SharePoint. If other people within your organization also use PnP Powershell then all the requests towards SharePoint are coming from the same identity and as a result you might hit throttling earlier.

**Reply by erwinvanhunen** [reply]
if you use -Credentials with the legacy version than you're connecting with your own identity, so then only your account gets throttled. (throttling is identity based)

**Reply by erwinvanhunen** [reply]
If you use -ClientId/-ClientSecret or -ClientId/Certificate in either version and more people use those Ids, then the same identity is used throughout the various scripts people run, and that can cause throttling earlier too.

**Reply by erwinvanhunen** [reply]
Another reason can be is that the farm your tenant is hosted on is very busy at times, which is unfortunately out of your control

**Reply by MartinHatch** [reply]
I'm doing -UseWebLogin

**Reply by erwinvanhunen** [reply]
okay, so that means you're using your own identity, e.g. it's your user account that gets throttled.

**Reply by erwinvanhunen** [reply]
We're going to check some things on this end. I will get back to you (most likely tomorrow).

**Reply by erwinvanhunen** [reply]
Short intermediate question:

Can you turn on logging before you start your script?

```powershell
Set-PnPTraceLog -On -Logfile c:\pathto\your\logfile.txt
<yourcmds>
Set-PnPTraceLog -Off
```

If the underlying SDKs encounter a throttling issue we add an entry to the log stating "CSOM request frequency exceeded usage limits. Sleeping for {0} seconds before retrying.". If we run into an exception we do not throw a 429 ourselves but a "MaximumRetryException: Maximum retry attemps {count} has been attempted"

**Reply by MartinHatch** [reply]
Yep I'm running it through again with the TraceLog set ... 

I was definitely seeing Exceptions thrown from ExecuteQuery telling the server response was HTTP 429 ..

But I'm seeing those "CSOM Request Frequency .. " entries in the tracelog too .. 

We will see (sometimes it takes 20-30 minutes before I get a 429)

**Reply by MartinHatch** [reply]
Ok ... and it seems even after posting the screenshot .. I didn't notice the extra little "ExecuteQuery" which had snuck in there ...
.. 
LOL

**Reply by erwinvanhunen** [reply]
too many threads here :-) But as I wrote in the other thread: we flipped the default behavior of the PnP.Framework to use the retry-after header now. As of PowerShell 0.3.14-nightly it will use that.

**Reply by AndersRask** [reply]
As this is still beta, I guess removing switch parameters is in order, but in general wouldn't it be best to avoid changes that will break code? :)

**Reply by AndersRask** [reply]
@MartinHatch using service principals instead of a normal user will be more resilient towards throttling. We often use multiple service principals to avoid throttling when migrating high volume content

**Comment by smajchrz** [comment]
Apologize for opening an old thread :)
What is the current state of handling throttling by PnP.Powershell? Is it using the Retry-After response header or incrementally changing wait time as described by @erwinvanhunen  in this discussion? I am asking as I see the following lines in the trace log:

```
[PnP.Framework]	[0]	[Information]	CSOM request frequency exceeded usage limits. Retry attempt 1. Sleeping for 500 milliseconds before retrying.	0ms   
[PnP.Framework]	[0]	[Information]	CSOM request frequency exceeded usage limits. Retry attempt 2. Sleeping for 1000 milliseconds before retrying.	0ms
```

**Reply by AndersRask** [reply]
PnP.Framework for most requests should use PnPFramework RetryHandler that handles throttling by using Retry-After. There is currently no support for RateLimit though (would like that very much!)

The exception that I am aware of is provisioning templates, that uses native SPWeb methods, so that is a "black box" with no reply when u get throttled

What kind of request did you get the above with? And what version? That looks like the "old" way of doing things, where they incremented wait time (not recommended) that should have been fixed in PnP.Framework

**Reply by smajchrz** [reply]
It is `1.11.0` version of the module.
As I can not see what command exactly is causing it, I can list cmdlets used in the script:
``` PowerShell
Get-PnPGroup
Get-PnPSiteCollectionAdmin
```
